[pytorch] make is_tracing scriptable

ppwwyyxx opened the pull request:
Summary: fix https://github.com/pytorch/pytorch/issues/47379

Test Plan: buck test mode/dev-nosan //caffe2/test:jit -- 'test_script_is_tracing'

Differential Revision: D25704315

albanD commented on test/jit/test_tracer.py:
```diff
@@ -1487,13 +1500,13 @@ def forward(self, x):
-        def test_index_put(target, indices, rhs):
-            target[indices] = rhs
-            return target
+    # def test_index_put_trace_with_view(self):
```
Should this be uncommented?

albanD commented on torch/jit/_trace.py:
```diff
@@ -976,6 +976,8 @@ def is_tracing():
     Returns ``True`` in tracing (if a function is called during the tracing of
     code with ``torch.jit.trace``) and ``False`` otherwise.
     """
+    if is_scripting():
```
Is that expected to always be true? We can never trace in script mode?

ppwwyyxx replied:
I'm not so sure. There is a "script in trace mode", but I'm not sure if it is or will be possible to do "trace in script mode" because trace requires extra inputs.

albanD replied:
We should make sure that this is the expected behavior?
cc @suo

SplitInfinity replied:
They should be mutually exclusive. Even though you can mix tracing and scripting (e.g. by tracing a module and then scripting its parent module or vice-versa) you can't do both at the same time.

albanD reviewed (commented):
Could you explain why the implementation of index_put is modified? Does it actually change the index_put function itself?

ppwwyyxx commented:
> Could you explain why the implementation of index_put is modified? Does it actually change the index_put function itself?

Those are unrelated. Original diff is here: D25704315, but looks like the diff export messed up the diff stack.

ppwwyyxx committed:
[pytorch] make is_tracing scriptable (#49853)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/49853

fix https://github.com/pytorch/pytorch/issues/47379

Test Plan: buck test mode/dev-nosan //caffe2/test:jit -- 'test_script_is_tracing'

Differential Revision: D25704315

fbshipit-source-id: 4c53b11368685270dc2e268aa1cf4ff96e712d02

albanD reviewed (changes requested):
The github export is not right. It includes quite a lot of changes not included in the diff :/

suo commented:
bouncing to @SplitInfinity as I think he has the most context
